library/Falcon/Handler/SES.php

```diff
@@ -0,0 +1,287 @@
+		admin_url('admin-post.php?action=bbsub')
+	) ?></p>
+<?php
+	}
+
+	public function send_mail( $users, Falcon_Message $message ) {
+		// Filter arguments to ensure both text and HTML are sent.
+		$callback = function ( $message_args ) use ( $message ) {
+			// Remove default HTML/text.
+			unset( $message_args['text'] );
+			unset( $message_args['html'] );
+
+			// Re-add as appropriate.
+			if ( $text = $message->get_text() ) {
+				$message_args['text'] = $text;
+			}
+			if ( $html = $message->get_html() ) {
+				$message_args['html'] = $html;
+			}
+			return $message_args;
+		};
+
+		add_filter( 'aws_ses_wp_mail_message_args', $callback );
+
+		$res = parent::send_mail( $users, $message );
+
+		remove_filter( 'aws_ses_wp_mail_message_args', $callback );
+
+		return $res;
+	}
+
+	public function handle_post() {
+		$input = file_get_contents( 'php://input' );
+		if ( empty( $input ) ) {
+			header( 'X-Fail: No input', true, 400 );
+			echo 'No input found.'; // intentionally not translated
+			return;
+		}
+
+		$data = json_decode( $input );
+		if ( ! $data && json_last_error() !== JSON_ERROR_NONE ) {
+			header( 'X-Fail: Bad input', true, 400 );
+			echo 'Could not decode input.'; // intentionall not translated
+			return;
+		}
+
+		// Determine the type of SNS notification.
+		$type = $_SERVER['HTTP_X_AMZ_SNS_MESSAGE_TYPE'] ?? 'no-header';
+
+		switch ( $type ) {
+			case 'SubscriptionConfirmation':
+				$this->handle_subscription_confirmation( $data );
+				break;
+
+			case 'Notification':
+				$this->handle_incoming_email( $data );
+				break;
+
+			default:
+				header( 'X-Fail: Unknown type', true, 400 );
+				printf( 'Could not parse type "%s"', $type );
+				break;
+		}
+	}
+
+	/**
+	 * Handle incoming subscription confirmation events from SES.
+	 *
+	 * We need to fetch the subscribe URL to confirm our subscription.
+	 *
+	 * @param stdClass $data Data object from handle_post()
+	 */
+	protected function handle_subscription_confirmation( $data ) {
+		// Send a GET request to the subscription confirmation URL.
+		$result = wp_remote_get( $data->SubscribeURL );
+		if ( wp_remote_retrieve_response_code( $result ) !== 200 ) {
+			header( 'X-Fail: Could not confirm', true, 500 );
+			echo 'Could not confirm subscription.';
+			return;
+		}
+
+		// Mark it as confirmed.
+		Falcon::update_option( 'bbsub_ses_topic_arn', $data->TopicArn );
+	}
+
+	/**
+	 * Handle incoming email events from SES.
+	 *
+	 * @link https://docs.aws.amazon.com/ses/latest/DeveloperGuide/receiving-email-notifications-contents.html
+	 *
+	 * @param stdClass $data Data object from handle_post()
+	 */
+	protected function handle_incoming_email( $data ) {
+		if ( $data->notificationType !== 'Received' ) {
+			// "For this type of notification, the value is always Received."
+			// https://docs.aws.amazon.com/ses/latest/DeveloperGuide/receiving-email-notifications-contents.html
+			return;
+		}
+
+		$reply = new Falcon_Reply();
```

Comment:
I'm not seeing us verify the request is actually from SNS, is there any way we can verify that? Could someone spam fake replies by just hitting this endpoint?

Reply:
No, as the To address includes a nonce which is verified.

We aren't verifying it's from SNS though; to do so, we'd pretty much need to include the AWS SDK, which isn't a hard requirement currently. Maybe we should optionally verify if it's loaded?

Reply:
Ahh ok didn't see the nonce, i think that should cover it then